pdfminer/utils.py

```diff
@@ -99,17 +99,45 @@ def apply_png_predictor(pred, colors, columns, bitspercomponent, data):
                 line2 += bytes((c,))
         elif ft == 3:
             # PNG average (UNTESTED)
+            a = 0
+            for (b,x) in zip(line0, line1):
+                x = (x + (a+b)//2) & 255  #((c + a + b) // 2) & 255
+                a=x
+                line2 += bytes((x,))
+        elif ft == 4:
+            #Paeth
+            """ 
+            Bytes in lines
+            c b  ---- line0
+            a x  ---- line1
+            """
+            
             c = 0
-            for (a, b) in zip(line0, line1):
-                c = ((c + a + b) // 2) & 255
-                line2 += bytes((c,))
+            a = 0
+            for (b, x) in zip(line0, line1):
+                x = (x+PaethPredictor(a,b,c)) & 255
+                c = b
+                a = x
+                line2 += bytes((x,))
```

Comment:
Where did you get the algorithm from? From the PDF reference? For future readers will be very pleased if you can add a url to it here.